src/ophyd_async/core/_providers.py

```diff
@@ -17,41 +20,105 @@ class DirectoryInfo:
 
     root: Path
     resource_dir: Path
-    prefix: Optional[str] = ""
-    suffix: Optional[str] = ""
+    filename: Optional[str] = ""
+    create_dir_depth: Optional[int] = 0
+
 
+class FilenameProvider(Protocol):
+
+    @abstractmethod
+    def __call__(self, prefix="", suffix="", device_name=None) -> str:
+        """Get a filename to use for output data"""
 
 class DirectoryProvider(Protocol):
+
+    _filename_provider: FilenameProvider
+    _prefix: Optional[str]
+    _suffix: Optional[str]
+
     @abstractmethod
-    def __call__(self) -> DirectoryInfo:
+    def __call__(self, device_name=None) -> DirectoryInfo:
         """Get the current directory to write files into"""
 
 
+class StaticFilenameProvider(FilenameProvider):
+    def __init__(self, filename):
+        self._static_filename = filename
+
+    def __call__(self, prefix="", suffix="", device_name=None) -> str:
+        return f'{prefix}{self._static_filename}{suffix}'
+
+
+class DeviceNameFilenameProvider(FilenameProvider):
+
+    def __call__(self, prefix="", suffix="", device_name=None) -> str:
+        return f'{prefix}{device_name}{suffix}'
+
+
+class UUIDFilenameProvider(FilenameProvider):
+
+    def __call__(self, prefix="", suffix="", device_name=None) -> str:
+        return f'{prefix}{uuid.uuid4()}{suffix}'
+
+
 class StaticDirectoryProvider(DirectoryProvider):
     def __init__(
         self,
+        filename_provider: FilenameProvider,
         directory_path: Union[str, Path],
         filename_prefix: str = "",
         filename_suffix: str = "",
         resource_dir: Path = Path("."),
+        create_dir_depth: int = 0,
     ) -> None:
+        self._filename_provider = filename_provider
+        self._directory_path = directory_path
         if isinstance(directory_path, str):
-            directory_path = Path(directory_path)
-        self._directory_info = DirectoryInfo(
-            root=directory_path,
-            resource_dir=resource_dir,
-            prefix=filename_prefix,
-            suffix=filename_suffix,
+            self._directory_path = Path(directory_path)
+        self._resource_dir = resource_dir
+        self._prefix = filename_prefix
+        self._suffix = filename_suffix
+        self._create_dir_depth = create_dir_depth
+
+    def __call__(self, device_name=None) -> DirectoryInfo:
+        return DirectoryInfo(
+            root=self._directory_path,
+            resource_dir=self._resource_dir,
+            filename=self._filename_provider(prefix=self._prefix, suffix=self._suffix, device_name=device_name),
+            create_dir_depth=self._create_dir_depth
         )
 
-    def __call__(self) -> DirectoryInfo:
-        return self._directory_info
+
+class YMDDirectoryProvider(DirectoryProvider):
+    def __init__(
+        self,
+        filename_provider: FilenameProvider,
+        directory_path: Union[str, Path],
+        filename_prefix: str = "",
+        filename_suffix: str = "",
+        create_dir_depth: int = 0,
```

Comment:
Probably want to default this to `-3` if you want Y/M/D directories to be made?

Reply:
I think that makes sense, will do.